### PR TITLE
Support for ADFS 640K (.adl) disk image files

### PR DESCRIPTION
**Pull request by robsonjb1**
This change enables support for the most common format of ADFS disk (640k .adl files)
These have 16 sectors per track rather than 10 (the norm for DFS).

This ADFS Welcome disk for the Master 128 can now be loaded. (This change is also of use to the 1770 DFS model which also includes the ADFS ROM).

I haven't also included the Master Welcome/Utilities disk. I can do this but was conscious of not wanting to bloat out the catalogue and list of bundled disk images.

### CODE REVIEW

**Comment by mattgodbolt** [comment]
Thanks for your patience on this! and for rebasing!! I'll try and get to this today or tomorrow!

**Comment by robsonjb1** [comment]
> Thanks for your patience on this! and for rebasing!! I'll try and get to this today or tomorrow!

Apologies I hadn't realised there were merge conflicts. There is also a new function (on the disk menu) to offer the user a download of the current drive 0 image ... (so you can persist changes to drives)

**Comment by mattgodbolt** [comment]
Oh that's awesome!

File: fdc.js [hunk 1]
```diff
@@ -146,7 +148,7 @@ export function BaseDisc(fdc, name, data, flusher) {
             case 6:
                 this.fdc.discFinishRead();
                 this.rsector++;
-                if (this.rsector === 10) this.rsector = 0;
+                if (this.rsector === (this.sectorsPerTrack)) this.rsector = 0;
```
Comment (mattgodbolt) [inline fdc.js]: super nitpick (and I'm surprised the `prettier` thing didn't pick up on this) - would you mind removing the extra parens here?

In fact, I may have broken the pre-commit hooks; can you run:
```
$ npx prettier --write .
```

(or similar) and see if it fixes up/complains :) thanks!

File: fdc.js [hunk 2]
```diff
@@ -78,6 +78,7 @@ export function BaseDisc(fdc, name, data, flusher) {
     if (data === null || data === undefined) throw new Error("Bad disc data");
     let nameDetails = utils.discImageSize(name);
     let isDsd = nameDetails.isDsd;
+    let sectorsPerTrack = nameDetails.isDoubleDensity ? 16 : 10;
```
Comment (mattgodbolt) [inline fdc.js:81]: Much nicer, thanks :)

File: index.html [hunk 3]
```diff
@@ -435,30 +436,30 @@ <h5>Handy key mappings</h5>
...
+                        <a href="#"><span class="name"></span></a> - <span class="description"></span>
+                    </li>
+                </ul>
+                To load a custom disc image, get an SSD, DSD or ADL file and load it below. Search the web, or check
+                somewhere like
+                <a href="http://www.bbcmicrogames.com/GettingStarted.html">here</a> for these. Be aware the
+                images are usually stored in a ZIP file, and you'll need to unzip first.
+                <div class="disc">
+                    <label>Load local SSD, DSD or ADL file:
+                        <input type="file" id="disc_load" accept=".ssd,.dsd,.adl,application/binary">
```
Comment (mattgodbolt) [inline index.html]: Thanks!

File: main.js [hunk 4]
```diff
@@ -1103,6 +1103,19 @@ $("#google-drive form").on("submit", function (e) {
...
+    var a = document.createElement("a");
+    document.body.appendChild(a);
+    a.style = "display: none";
+
+    var blob = new Blob([processor.fdc.drives[0].data], {type: 'application/octet-stream'}),
+    url = window.URL.createObjectURL(blob);
+    a.href = url;
+    a.download = processor.fdc.drives[0].name;
+    a.click();
+    window.URL.revokeObjectURL(url);
```
Comment (mattgodbolt) [inline main.js:1116]: Super cool! Thanks so much! TIL!

File: models.js [hunk 5]
```diff
@@ -70,7 +70,7 @@ export const allModels = [
         false
     ),
     new Model(
-        "BBC B with Teletext",
+        "BBC B (with Teletext)",
```
Comment (mattgodbolt) [inline models.js:73]: Thanks too - this looks better!

File: fdc.js [hunk 6]
```diff
@@ -174,14 +176,14 @@ export function BaseDisc(fdc, name, data, flusher) {
...
-        this.seekOffset = track * 10 * 256;
+        this.seekOffset = track * this.sectorsPerTrack * 256;
         if (this.isDsd) this.seekOffset <<= 1;
         const oldTrack = this.track;
         this.track = track;
         return this.track - oldTrack;
     };
-    BaseDisc.prototype.check = (track, side, density) => {
-        if (this.track !== track || density || (side && !this.isDsd)) {
+    BaseDisc.prototype.check = (track, side) => {
```
Comment (mattgodbolt) [inline fdc.js:185]: unless I'm mistaken are there calls to `check()` somewhere passing `density` ? I can't see that you removed the `density` param, though js is forgiving I'd prefer to remove that if possible.

**Review by mattgodbolt — COMMENTED** [review]
Basically looks amazing! couple super minor points only!

File: fdc.js (hunk 6 above)
Comment (robsonjb1) [reply fdc.js:185]: Hi there, yes the `check()` seemed to be using `density` just in order to create an error on non single-density disks. I've tidied up the various calls to it now it isn't needed to be passed in.

File: fdc.js (hunk 1 above)
Comment (robsonjb1) [reply fdc.js]: Yes I wasn't sure at what point all the pretty stuff was going to happen, didn't seem to launch on pre-commit so have run it manually.

File: fdc.js (hunk 6 above)
Comment (mattgodbolt) [reply fdc.js:185]: awesome thanks!

**Review by mattgodbolt — APPROVED** [review]
Awesome stuff! Merging now: thanks so much!

**Pull request merged**